### PR TITLE
use spf13/cobra instead of urfave/cli

### LINKED ISSUE #95: consolidate binaries
Should we consolidate `counter` and `dummy` into the `abci-cli` ? So something like `abci-cli counter` and `abci-cli dummy` ?

### LINKED ISSUE #101: use spf13/cobra instead of urfave/cli

### PR DESCRIPTION
**Pull request by zramsay**
- closes #95, #101

### CODE REVIEW

File: cmd/abci-cli/abci-cli.go [hunk 1]
```diff
@@ -38,149 +43,181 @@ var client abcicli.Client
...
+	Short: "Run a batch of abci commands against an application",
+	Long:  "",
+	Run:   cmdBatch,
+}
+
+var consoleCmd = &cobra.Command{
+	Use:       "console",
+	Short:     "Start an interactive abci console for multiple commands",
+	Long:      "",
+	ValidArgs: []string{"batch", "echo", "info", "set_option", "deliver_tx", "check_tx", "commit", "query"},
```
Comment (zramsay) [inline cmd/abci-cli/abci-cli.go]: replaces https://github.com/tendermint/abci/pull/117/files#diff-4a39505c1ed5bc7bab8a6b542891e13bL176

File: cmd/abci-cli/abci-cli.go [hunk 2]
```diff
@@ -196,164 +233,152 @@ func persistentArgs(line []byte) []string {
...
 //--------------------------------------------------------------------------------
 
-func cmdBatch(app *cli.App, c *cli.Context) error {
+func cmdBatch(cmd *cobra.Command, args []string) {
 	bufReader := bufio.NewReader(os.Stdin)
 	for {
 		line, more, err := bufReader.ReadLine()
 		if more {
-			return errors.New("Input line is too long")
+			ifExit(errors.New("Input line is too long"))
```
Comment (ebuchman) [inline cmd/abci-cli/abci-cli.go]: We should only use `RunE` and return errors here.

File: cmd/abci-cli/abci-cli.go [hunk 3]
```diff
@@ -196,164 +233,152 @@ func persistentArgs(line []byte) []string {
...
 			continue
 		} else if err != nil {
-			return err
+			ifExit(err)
 		}
 
-		args := persistentArgs(line)
-		app.Run(args) //cli prints error within its func call
+		pArgs := persistentArgs(line)
+		out, err := exec.Command(pArgs[0], pArgs[1:]...).Output()
```
Comment (ebuchman) [inline cmd/abci-cli/abci-cli.go]: shelling it out? would be better if we could keep in process

File: cmd/abci-cli/abci-cli.go [hunk 4]
```diff
@@ -196,164 +233,152 @@ func persistentArgs(line []byte) []string {
...
-	return nil
 }
 
 // Validate a tx
-func cmdCheckTx(c *cli.Context) error {
-	args := c.Args()
+func cmdCheckTx(cmd *cobra.Command, args []string) {
 	if len(args) != 1 {
-		return errors.New("Command check_tx takes 1 argument")
+		ifExit(errors.New("Command check_tx takes only 1 argument"))
```
Comment (ebuchman) [inline cmd/abci-cli/abci-cli.go]: no need for "only"

File: cmd/abci-cli/abci-cli.go [hunk 5]
```diff
@@ -363,17 +388,70 @@ func cmdQuery(c *cli.Context) error {
...
+	app := counter.NewCounterApplication(serial)
 
-	verbose := c.GlobalBool("verbose")
+	logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+
+	// Start the listener
+	srv, err := server.NewServer(addrC, abci, app)
+	if err != nil {
+		logger.Error(err.Error())
+		os.Exit(1)
```
Comment (ebuchman) [inline cmd/abci-cli/abci-cli.go]: return err

File: cmd/abci-cli/abci-cli.go [hunk 6]
```diff
@@ -426,3 +499,10 @@ func stringOrHexToBytes(s string) ([]byte, error) {
 
 	return []byte(s[1 : len(s)-1]), nil
 }
+
+func ifExit(err error) {
```
Comment (ebuchman) [inline cmd/abci-cli/abci-cli.go]: remove this and all use of `os.Exit` in functions

**Pull request merged**